File: CCWCenterline/src/utils/overlay.js
// Printing centerline values onto the RCU screens, in the machine's own boxes.
//
// The screenshots have one machine's numbers baked into them (a 14-head unit
// running POTATO CHIPS at 90.0 g). To show a different customer's settings each
// value box is erased and the new value drawn in the RCU's own blue, so the
// result reads as the screen that customer should be looking at. That is the
// point of the document: an operator holds the page against the live machine
// and compares field by field.
//
// Erasing is the fiddly half. A flat fill is wrong — the field faces carry a
// vertical gradient, so one sampled colour leaves an obvious patch and the page
// looks doctored. Each row is repainted with the median of its OWN non-text
// pixels, which follows the gradient and works equally on the grey buttons and
// the green list panels.
//
// Port of tools/overlay.py, same approach and same output.

export const RCU_BLUE = '#0000cc';

/** Is this pixel part of the printed value rather than the field behind it? */
const isValueInk = (r, g, b) => r < 140 && g < 140 && b > r + 30;

/**
 * Erase one value box, leaving the field looking untouched.
 *
 * Works on an ImageData in place. `box` is [x0, y0, x1, y1] inclusive.
 */
export function repaintBox(imageData, box) {
  const [x0, y0, x1, y1] = box;
  const { data, width } = imageData;
  const row = [];

  for (let y = Math.max(0, y0); y <= Math.min(imageData.height - 1, y1); y += 1) {
    row.length = 0;
    for (let x = Math.max(0, x0); x <= Math.min(width - 1, x1); x += 1) {
      const i = (y * width + x) * 4;
      const r = data[i];
      const g = data[i + 1];
      const b = data[i + 2];
      if (!isValueInk(r, g, b)) row.push([r, g, b]);
    }
    if (!row.length) continue;
    // Median by luminance: robust to the odd antialiased edge pixel that the
    // ink test lets through, where a mean would drag the whole row darker.
    row.sort((p, q) => (p[0] + p[1] + p[2]) - (q[0] + q[1] + q[2]));
    const [mr, mg, mb] = row[row.length >> 1];
    for (let x = Math.max(0, x0); x <= Math.min(width - 1, x1); x += 1) {
      const i = (y * width + x) * 4;
      data[i] = mr;
      data[i + 1] = mg;
      data[i + 2] = mb;
    }
  }
}

/** The text to print for a field, with its unit appended if not already there. */
export function displayValue(value, unit) {
  const text = String(value ?? '').trim();
  if (!text) return '';
  if (!unit) return text;
  return text.endsWith(unit) ? text : text + unit;
}

/**
 * Draw a screen with the centerline's values in place of the baked-in ones.
 *
 * `image` is a loaded HTMLImageElement of the 800x600 screenshot. Returns a
 * canvas; fields with no value are left exactly as the screenshot has them,
 * which is deliberate — an untouched field is visibly the sample machine's,
 * and a blank one would read as "set this to nothing".
 */
export function renderScreen(image, fields, values) {
  const canvas = document.createElement('canvas');
  canvas.width = image.naturalWidth || image.width;
  canvas.height = image.naturalHeight || image.height;
  const ctx = canvas.getContext('2d');
  ctx.drawImage(image, 0, 0);

  const written = [];
  const wanted = fields.filter((f) => displayValue(values?.[f.key], f.unit));
  if (!wanted.length) return { canvas, written };

  const frame = ctx.getImageData(0, 0, canvas.width, canvas.height);
  for (const field of wanted) repaintBox(frame, field.box);
  ctx.putImageData(frame, 0, 0);

  ctx.textAlign = 'center';
  ctx.textBaseline = 'middle';
  ctx.fillStyle = RCU_BLUE;
  ctx.font = 'bold 15px Helvetica, Arial, sans-serif';
  for (const field of wanted) {
    const [x0, y0, x1, y1] = field.box;
    const text = displayValue(values[field.key], field.unit);
    ctx.fillText(text, (x0 + x1) / 2, (y0 + y1) / 2, x1 - x0 - 4);
    written.push({ label: field.label, text });
  }
  return { canvas, written };
}

/** Load an image URL, resolving only once it can actually be drawn. */
export const loadImage = (src) =>
  new Promise((resolve, reject) => {
    const img = new Image();
    img.crossOrigin = 'anonymous';
    img.onload = () => resolve(img);
    img.onerror = () => reject(new Error(`could not load ${src}`));
    img.src = src;
  });

File: CCWCenterline/tools/overlay.py
#!/usr/bin/env python3
"""Print centerline values onto the RCU screens, in the machine's own boxes.

The screenshots have one customer's numbers baked into them (POTATO CHIPS at
90.0g). To show somebody else's settings, each value box is repainted in the
field's own background colour and the new value drawn in the RCU's blue. The
result reads as the screen that customer should be looking at, which is the
point: an operator can hold it against the live machine and compare.

Usage: python3 tools/overlay.py <screens-dir> <values.json> <outdir>
"""
import json, os, sys
from PIL import Image, ImageDraw, ImageFont

HERE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
RCU_BLUE = (0, 0, 204)


def _font(size):
    for path in ('/System/Library/Fonts/Supplemental/Arial Bold.ttf',
                 '/System/Library/Fonts/Helvetica.ttc'):
        if os.path.exists(path):
            return ImageFont.truetype(path, size)
    return ImageFont.load_default()


def _repaint(image, box):
    """Erase a value, leaving the field looking untouched.

    A flat fill is wrong: these buttons have a vertical gradient across their
    face, so a single sampled colour leaves a visible patch. Each row is
    repainted with the median of its OWN non-text pixels instead, which follows
    the gradient and works just as well on the green list panels as on grey.
    """
    x0, y0, x1, y1 = box
    px = image.load()
    for y in range(y0, y1 + 1):
        row = []
        for x in range(x0, x1 + 1):
            r, g, b = px[x, y]
            # The value text is dark and blue; anything else is field.
            if not (r < 140 and g < 140 and b > r + 30):
                row.append((r, g, b))
        if not row:
            continue
        row.sort(key=lambda c: c[0] + c[1] + c[2])
        mid = row[len(row) // 2]
        for x in range(x0, x1 + 1):
            px[x, y] = mid


def render(screen_path, fields, values, out_path):
    im = Image.open(screen_path).convert('RGB')
    draw = ImageDraw.Draw(im)
    written = []
    for field in fields:
        value = values.get(field['key'])
        if value is None:
            continue
        box = field['box']
        _repaint(im, box)
        text = str(value)
        if field.get('unit') and not text.endswith(field['unit']):
            text += field['unit']
        font = _font(15)
        w = draw.textlength(text, font=font)
        cx = (box[0] + box[2]) / 2
        cy = (box[1] + box[3]) / 2
        draw.text((cx - w / 2, cy - 9), text, fill=RCU_BLUE, font=font)
        written.append((field['label'], text))
    im.save(out_path)
    return written


def main(screens_dir, values_path, outdir):
    spec = json.load(open(os.path.join(HERE, 'data', 'rcu-fields.json')))
    values = json.load(open(values_path))
    os.makedirs(outdir, exist_ok=True)
    for slug, screen in spec['screens'].items():
        vals = values.get(slug, {})
        if not vals:
            continue
        src = os.path.join(screens_dir, slug + '.jpg')
        if not os.path.exists(src):
            print('missing screen:', src)
            continue
        out = os.path.join(outdir, slug + '.png')
        written = render(src, screen['fields'], vals, out)
        print('%s -> %d values' % (slug, len(written)))
        for label, text in written:
            print('    %-34s %s' % (label, text))


if __name__ == '__main__':
    if len(sys.argv) != 4:
        raise SystemExit(__doc__)
    main(*sys.argv[1:])
